packages/service-utils/src/core/usageLimit/index.ts

```diff
@@ -0,0 +1,36 @@
+import { ApiKeyMetadata, CoreServiceConfig } from "../api";
+
+import { UsageLimitResult } from "./types";
+
+export async function usageLimit(
+  apiKeyMeta: ApiKeyMetadata,
+  serviceConfig: CoreServiceConfig,
+): Promise<UsageLimitResult> {
+  const { limits, usage } = apiKeyMeta;
+  const { serviceScope } = serviceConfig;
+
+  const limit = limits[serviceScope];
+
+  if (!usage || !(serviceScope in usage) || limit === undefined) {
+    // No usage limit is provided. Assume the request is not limited.
+    return {
+      usageLimited: false,
+    };
+  }
+
+  if (
+    serviceScope === "storage" &&
+    (usage.storage?.sumFileSizeBytes || 0) > limit
+  ) {
+    return {
+      usageLimited: true,
+      status: 402,
```

Comment:
Interesting, I didn't know 402 existed but I see it's [experimental](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/402). No strong opinion. [403 Forbidden](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403) would be a reasonable alternative.